File: latticeai/tools/documents.py
"""Document generation (docx/xlsx/pptx/pdf) and extraction (read_document)."""

from __future__ import annotations

from pathlib import Path
from typing import Any, Dict, Optional, Tuple

from latticeai.tools import (
    _SUPPORTED_READ_EXTENSIONS,
    DOCUMENT_MAX_READ_BYTES,
    DOCUMENT_OUTPUT_DIR,
    PDF_OUTPUT_DIR,
    PRESENTATION_OUTPUT_DIR,
    SPREADSHEET_OUTPUT_DIR,
    ToolError,
)


def _safe_filename(name: str, suffix: str) -> str:
    base = Path(name or f"artifact{suffix}").name
    if not base.lower().endswith(suffix):
        base += suffix
    safe = "".join(ch if ch.isalnum() or ch in ("-", "_", ".", " ") else "_" for ch in base).strip()
    return safe or f"artifact{suffix}"


#: tool name → (output directory, enforced suffix). The creators below are the
#: only writers, so this table is the single source of truth for "where does
#: this document actually land?".
_DOCUMENT_TOOL_TARGETS: Dict[str, Tuple[str, str]] = {
    "create_docx": (DOCUMENT_OUTPUT_DIR, ".docx"),
    "create_xlsx": (SPREADSHEET_OUTPUT_DIR, ".xlsx"),
    "create_pptx": (PRESENTATION_OUTPUT_DIR, ".pptx"),
    "create_pdf": (PDF_OUTPUT_DIR, ".pdf"),
}


def document_output_target(tool_name: str, filename: str) -> Optional[str]:
    """Workspace-relative path ``tool_name`` will write ``filename`` to.

    Governance needs this: the creators sanitize the caller's ``filename``
    into their own output directory, so an "does the target already exist?"
    check against the raw argument looks at a path nothing ever writes — and
    the fail-closed overwrite guard would silently never fire.

    Returns ``None`` for tools that write wherever the caller points them, so
    callers fall back to the raw argument.
    """
    entry = _DOCUMENT_TOOL_TARGETS.get(tool_name)
    if entry is None:
        return None
    output_dir, suffix = entry
    return f"{output_dir}/{_safe_filename(filename, suffix)}"


def _body_to_str(body) -> str:
    if isinstance(body, list):
        return "\n\n".join(str(item) for item in body)
    return str(body or "")


def read_document(path: str) -> Dict[str, Any]:
    """Extract text from PDF, DOCX, XLSX, PPTX, TXT, MD, CSV files."""
    target = Path(path).expanduser().resolve()
    if not target.exists():
        raise ToolError(f"파일이 없습니다: {path}")
    if not target.is_file():
        raise ToolError(f"파일이 아닙니다: {path}")
    if target.stat().st_size > DOCUMENT_MAX_READ_BYTES:
        raise ToolError(f"파일이 너무 큽니다 ({target.stat().st_size:,} bytes).")

    ext = target.suffix.lower()
    if ext not in _SUPPORTED_READ_EXTENSIONS:
        raise ToolError(f"지원하지 않는 형식입니다: {ext}. 지원: {', '.join(_SUPPORTED_READ_EXTENSIONS)}")

    text = ""
    meta: Dict[str, Any] = {"path": str(target), "ext": ext}

    if ext == ".pdf":
        try:
            import pdfplumber
            with pdfplumber.open(str(target)) as pdf:
                meta["pages"] = len(pdf.pages)
                text = "\n\n".join(
                    (p.extract_text() or "") for p in pdf.pages
                ).strip()
        except Exception as exc:
            raise ToolError(f"PDF 읽기 실패: {exc}") from exc

    elif ext == ".docx":
        try:
            from docx import Document
            doc = Document(str(target))
            paragraphs = [p.text for p in doc.paragraphs if p.text.strip()]
            text = "\n\n".join(paragraphs)
            meta["paragraphs"] = len(paragraphs)
        except Exception as exc:
            raise ToolError(f"DOCX 읽기 실패: {exc}") from exc

    elif ext == ".xlsx":
        try:
            from openpyxl import load_workbook
            wb = load_workbook(str(target), data_only=True)
            rows_all = []
            for ws in wb.worksheets:
                rows_all.append(f"[Sheet: {ws.title}]")
                for row in ws.iter_rows(values_only=True):
                    cells = [str(c) if c is not None else "" for c in row]
                    rows_all.append("\t".join(cells))
            text = "\n".join(rows_all)
            meta["sheets"] = len(wb.worksheets)
        except Exception as exc:
            raise ToolError(f"XLSX 읽기 실패: {exc}") from exc

    elif ext == ".pptx":
        try:
            from pptx import Presentation
            prs = Presentation(str(target))
            slides_text = []
            for i, slide in enumerate(prs.slides, 1):
                parts = []
                for shape in slide.shapes:
                    if shape.has_text_frame:
                        parts.append(shape.text_frame.text)
                slides_text.append(f"[Slide {i}]\n" + "\n".join(parts))
            text = "\n\n".join(slides_text)
            meta["slides"] = len(prs.slides)
        except Exception as exc:
            raise ToolError(f"PPTX 읽기 실패: {exc}") from exc

    elif ext in {".txt", ".md", ".csv"}:
        try:
            text = target.read_text(encoding="utf-8", errors="replace")
        except Exception as exc:
            raise ToolError(f"파일 읽기 실패: {exc}") from exc

    meta["chars"] = len(text)
    meta["preview"] = text[:500]
    meta["content"] = text[:50_000]   # 50K char cap for context
    return meta

File: rust/lattice-agent/src/tools/documents.rs
//! `document_output_target` — where a document creator actually writes.
//!
//! A port of `latticeai.tools.documents`' target table, needed by exactly one
//! caller: the fail-closed overwrite guard. The creators sanitise the caller's
//! `filename` into their own output directory, so an "does the target already
//! exist?" check against the raw argument inspects a path nothing ever writes
//! and the guard silently never fires. That was the bug the Python version
//! exists to prevent, and a port without it would reintroduce it.

/// Tool name → (output directory, enforced suffix).
const DOCUMENT_TOOL_TARGETS: [(&str, &str, &str); 4] = [
    ("create_docx", "generated_documents", ".docx"),
    ("create_pdf", "generated_pdfs", ".pdf"),
    ("create_pptx", "generated_presentations", ".pptx"),
    ("create_xlsx", "generated_spreadsheets", ".xlsx"),
];

/// `_safe_filename`: basename, enforced suffix, and every other character
/// replaced by `_`.
fn safe_filename(name: &str, suffix: &str) -> String {
    let fallback = format!("artifact{suffix}");
    let source = if name.is_empty() { &fallback } else { name };
    let mut base = crate::kernel::transcript::path_name(source).to_string();
    if base.is_empty() {
        base = fallback.clone();
    }
    if !base.to_lowercase().ends_with(suffix) {
        base.push_str(suffix);
    }
    let safe: String = base
        .chars()
        .map(|character| {
            if character.is_alphanumeric() || matches!(character, '-' | '_' | '.' | ' ') {
                character
            } else {
                '_'
            }
        })
        .collect();
    let trimmed = safe.trim().to_string();
    if trimmed.is_empty() {
        fallback
    } else {
        trimmed
    }
}

/// The workspace-relative path `tool_name` will write `filename` to, or `None`
/// for tools that write wherever the caller points them.
pub fn document_output_target(tool_name: &str, filename: &str) -> Option<String> {
    DOCUMENT_TOOL_TARGETS
        .iter()
        .find(|(name, _, _)| *name == tool_name)
        .map(|(_, directory, suffix)| format!("{directory}/{}", safe_filename(filename, suffix)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn each_creator_lands_in_its_own_directory_with_its_own_suffix() {
        assert_eq!(
            document_output_target("create_docx", "report"),
            Some("generated_documents/report.docx".into())
        );
        assert_eq!(
            document_output_target("create_pdf", "a/b/notes.pdf"),
            Some("generated_pdfs/notes.pdf".into())
        );
        assert_eq!(
            document_output_target("create_pptx", "deck.PPTX"),
            Some("generated_presentations/deck.PPTX".into())
        );
        assert_eq!(
            document_output_target("create_xlsx", ""),
            Some("generated_spreadsheets/artifact.xlsx".into())
        );
    }

    #[test]
    fn tools_outside_the_table_write_where_they_are_told() {
        for tool in ["write_file", "edit_file", "create_web_project", "unknown"] {
            assert_eq!(document_output_target(tool, "a.md"), None, "{tool}");
        }
    }

    #[test]
    fn unsafe_characters_are_replaced_not_rejected() {
        assert_eq!(
            document_output_target("create_docx", "../../etc/pa$$wd"),
            Some("generated_documents/pa__wd.docx".into()),
            "the basename is taken first, so traversal cannot survive"
        );
        assert_eq!(
            document_output_target("create_docx", "보고서"),
            Some("generated_documents/보고서.docx".into()),
            "alphanumeric is Unicode-wide, as `str.isalnum` is"
        );
    }
}
